backends/graphics/gles-custom/gles-custom.cpp

```diff
@@ -0,0 +1,703 @@
+ *
+ */
+
+// Needed so we don't get errors when including the Raspberry Pi headers
+#define FORBIDDEN_SYMBOL_ALLOW_ALL
+#include "backends/graphics/gles-custom/gles-custom.h"
+#include "common/textconsole.h"
+#include "common/config-manager.h"
+#ifdef USE_OSD
+#include "common/translation.h"
+#endif
+
+#include <GLES/gl.h>
+#include <EGL/egl.h>
+
+#ifdef USE_GLES_RPI
+#include <bcm_host.h>
+DISPMANX_ELEMENT_HANDLE_T dispman_element;
+DISPMANX_DISPLAY_HANDLE_T dispman_display;
+DISPMANX_UPDATE_HANDLE_T dispman_update;
+static EGL_DISPMANX_WINDOW_T nativewindow;
+#endif
+
+#ifdef USE_GLES_FBDEV
+// Includes for framebuffer size retrieval
+#include <sys/ioctl.h>
+#include <linux/fb.h>
+#include <linux/vt.h>
+#include <fcntl.h>
+#include <unistd.h>
+struct fbdev_window nativewindow;
+#endif
+
+#ifdef USE_GLES_KMS
+// This hacky define is needed since there is an struct member called "virtual" in xf86drm.h
+#define virtual __virtual
+#include <xf86drm.h>
+#undef virtual
+
+#include <xf86drmMode.h>
+#include <gbm.h>
+
+static struct {
+	struct gbm_device *dev;
+	struct gbm_surface *surface;
+} gbm;
+
+static struct {
+	int fd;
+	drmModeModeInfo *mode;
+	uint32_t crtc_id;
+	uint32_t connector_id;
+} drm;
+
+struct drm_fb {
+	struct gbm_bo *bo;
+	uint32_t fb_id;
+};
+
+struct gbm_bo *bo;
+struct drm_fb *fb;
+fd_set fds;
+
+void page_flip_handler(int fd, unsigned int frame,
+		  unsigned int sec, unsigned int usec, void *data)
+{
+	int *waiting_for_flip = (int*)data;
+	*waiting_for_flip = 0;
+}
+
+static drmEventContext evctx;
+
+void drm_fb_destroy_callback(struct gbm_bo *bob, void *data);
+
+struct drm_fb * drm_fb_get_from_bo(struct gbm_bo *bob)
+{
+	struct drm_fb *fbu = (drm_fb*)gbm_bo_get_user_data(bob);
+	uint32_t width, height, stride, handle;
+
+	if (fbu) {
+		return fbu;
+	}
+	
+	fbu = (drm_fb*)calloc(1, sizeof *fbu);
+	fbu->bo = bob;
+
+	width = gbm_bo_get_width(bob);
+	height = gbm_bo_get_height(bob);
+	stride = gbm_bo_get_stride(bob);
+	handle = gbm_bo_get_handle(bob).u32;
+
+	if (drmModeAddFB(drm.fd, width, height, 24, 32, stride, handle, &fbu->fb_id)) {
+		debug("Could not add drm framebuffer\n");
+		free(fbu);
+		return NULL;
+	}
+
+	// We used to pass the destroy callback function here. Now it's done manually in deinit_egl()
+	gbm_bo_set_user_data(bob, fbu, NULL);
+	return fbu;
+}
+
+void DRM_PageFlip(void){
+	struct gbm_bo *next_bo;
+	int waiting_for_flip = 1;
+
+	next_bo = gbm_surface_lock_front_buffer(gbm.surface);
+	fb = drm_fb_get_from_bo(next_bo);
+
+	if (drmModePageFlip(drm.fd, drm.crtc_id, fb->fb_id,
+			DRM_MODE_PAGE_FLIP_EVENT, &waiting_for_flip)){
+		debug ("Failed to queue pageflip\n");
+		return;
+	}
+
+	while (waiting_for_flip) {
+		FD_ZERO(&fds);
+		
+		FD_SET(0, &fds);
+		FD_SET(drm.fd, &fds);
+
+		select(drm.fd+1, &fds, NULL, NULL, NULL);
+		
+		drmHandleEvent(drm.fd, &evctx);
+	}
+	
+	// release last buffer to render on again
+	gbm_surface_release_buffer(gbm.surface, bo);
+	bo = next_bo;
+}
+
+int init_drm(void)
```

Comment:
Once again, wrong formatting, bracket should stick to previous line, and why C syntax?